CARDS-1401: Creating forms via API ignores the maxPerSubject setting

IntegralProgrammer opened the pull request:
This PR implements CARDS-1401 by adding a _commit validator_ to ensure that commits of `cards:Form` nodes conform to the `maxPerSubject` constraint defined by their associated Questionnaire.

To test:

1. Build this `CARDS-1401-v2` branch with `mvn clean install`.
2. Start CARDS in `cards4proms` mode - `./start_cards.sh --dev --project cards4proms`
3. Create a new _Patient information_ Form along with its associated _Patient_ Subject
4. In a new browser tab, visit `http://localhost:8080/bin/browser.html`
5. _Check out_ the `Questionnaires/Patient information` node
6. Change the `Questionnaires/Patient information` `maxPerSubject` property from `1` to `10`
7. Using the CARDS Web UI, start the process of creating a new _Patient information_ Form for the previously created patient but before clicking _Create Form_, go to the browser tab with Composum open and change the `maxPerSubject` property of `/Questionnaires/Patient information` back to `1`.
8. Go back to the browser tab with the CARDS Web UI and click _Create Form_
9. Form creation should fail with the message `New form request failed with error code 409: Conflict`

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Create a boilerplate cards:Form node validator

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Add to the boilerplate the ability to read arbitrary resources in the JCR

Add an example of reading the maxPerSubject property of the "/Questionnaires/Patient information" node

sofi2002sofi committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

add calculation of existing forms per subject

sofi2002sofi committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

fix according to comments

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Remove debug logging statements

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Remove debug logging statements

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

We have already obtained the Questionnaire Resource by UUID therefore we do not need to re-obtain the UUID from the Questionnaire Resource since we already have the original UUID that was used in the first place to obtain the Questionnaire Resource

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Remove END constant and simply use "'" for the 3 times that it is needed

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Remove all remaining debug logging

IntegralProgrammer committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting

Improve commenting and JavaDoc

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/MaxFormsOfTypePerSubjectValidator.java:
```diff
@@ -0,0 +1,168 @@
+    public void propertyDeleted(PropertyState before) throws CommitFailedException
+    {
+    }
+
```
I think all these empty methods can be removed if this class extends DefaultValidator, isn't it so?

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/175764616-899d3c7b-8585-4a86-baa1-ba26a92e3f21.png)

sdumitriu commented on modules/data-entry/src/main/features/feature.json:
```diff
@@ -77,6 +77,11 @@
+    },
+    "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~maxFormsOfTypePerSubjectValidator":{
+      "user.mapping":[
+        "io.uhndata.cards.dataentry:maxFormsOfTypePerSubjectValidator=[sling-readall]"
```
This could be added in the array above, no need for a new configuration. But then the configuration name would need to be changed, since it's no longer just about `computedAnswers`.

sdumitriu committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting
Use DefaultValidator as a base class to avoid reimplementing many empty methods

sdumitriu committed:
CARDS-1401: Creating forms via API ignores the maxPerSubject setting
Use only one user mapping configuration

sdumitriu commented:
One thing to note: since this uses a query to count how many forms are already present, this is not yet a complete solution since the query depends on a successful lucene reindex. Since the index is asynchronous, there is a short-ish (few seconds) interval in which multiple forms can still be created. This isn't a big problem for the UI, but a CSV import may still create dozens or hundreds of forms for the same subject.
